src/openssl/digest/digest.cr

```diff
@@ -0,0 +1,88 @@
+require "../lib_crypto"
+require "./digest_base"
+
+module OpenSSL
+  class Digest
+    class DigestError < OpenSSL::Error; end
+
+    include DigestBase
+
+    getter name
+
+    def initialize(@name, @ctx : LibCrypto::EVP_MD_CTX)
+      raise DigestError.new("Invalid EVP_MD_CTX") unless @ctx
+    end
+
+    protected def self.create_evp_mt_ctx(name)
+      md = LibCrypto.evp_get_digestbyname(name)
+      unless md
+        oid = LibCrypto.obj_txt2obj(name, 0)
+        md = LibCrypto.evp_get_digestbyname(LibCrypto.obj_nid2sn(LibCrypto.obj_obj2nid(oid)))
+        LibCrypto.asn1_object_free(oid)
+      end
+      unless md
+        raise "Unsupported digest algoritm: #{name}"
+      end
+      ctx = LibCrypto.evp_md_ctx_create
+      unless ctx
+        raise OpenSSL::Digest::DigestError.new "Digest initialization failed."
+      end
+      if LibCrypto.evp_digestinit_ex(ctx, md, nil) != 1
+        raise OpenSSL::Digest::DigestError.new "Digest initialization failed."
+      end
+      ctx
+    end
+
+    def self.new(name)
+      new(name, create_evp_mt_ctx(name))
+    end
+
+    def finalize
+      LibCrypto.evp_md_ctx_destroy(self)
+    end
+
+    def clone
+      ctx = LibCrypto.evp_md_ctx_create
+      if LibCrypto.evp_md_ctx_copy(ctx, @ctx) == 0
+        LibCrypto.evp_md_ctx_destroy(ctx)
+        raise DigestError.new("Unable to clone digest")
+      end
+      Digest.new(@name, ctx)
+    end
+
+    def reset
+      if LibCrypto.evp_digestinit_ex(self, to_unsafe_md, nil) != 1
+        raise DigestError.new "Digest initialization failed."
+      end
+      self
+    end
+
+    def update(data : String | Slice)
+      LibCrypto.evp_digestupdate(self, data, data.size)
+      self
+    end
+
+    protected def finish
+      size = digest_size
+      data = Slice(UInt8).new(size)
+      LibCrypto.evp_digestfinal_ex(@ctx, data, nil)
+      data
```

Comment:
Why would that be faster?  The allocation code seems almost identical.  `to_slice` calls `Slice.new`.